src/worker/swirl/runner.rs

```diff
@@ -26,155 +23,153 @@ fn runnable<J: BackgroundJob>(ctx: &J::Context, payload: serde_json::Value) -> a
     ///
-    /// This function is intended for use in tests. If any jobs have failed, it
-    /// will return `swirl::JobsFailed` with the number of jobs that failed.
-    ///
-    /// If any other unexpected errors occurred, such as panicked worker threads
-    /// or an error loading the job count from the database, an opaque error
-    /// will be returned.
-    // FIXME: Only public for `src/tests/util/test_app.rs`
+    /// This function is intended for use in tests.
     pub fn check_for_failed_jobs(&self) -> anyhow::Result<()> {
-        self.wait_for_jobs()?;
         let failed_jobs = storage::failed_job_count(&mut *self.connection()?)?;
         if failed_jobs == 0 {
             Ok(())
         } else {
             Err(anyhow!("{failed_jobs} jobs failed"))
         }
     }
+}
 
-    fn wait_for_jobs(&self) -> anyhow::Result<()> {
-        self.thread_pool.join();
-        let panic_count = self.thread_pool.panic_count();
-        if panic_count == 0 {
-            Ok(())
-        } else {
-            Err(anyhow!("{panic_count} threads panicked"))
-        }
+pub struct RunningRunner {
+    handles: Vec<JoinHandle<()>>,
+}
+
+impl RunningRunner {
+    /// Wait for all background workers to shut down.
+    pub fn wait_for_shutdown(self) {
+        self.handles.into_iter().for_each(|handle| {
+            if let Err(error) = handle.join() {
+                let error = try_to_extract_panic_info(&error);
+                warn!(%error, "Background worker thread panicked");
+            }
+        })
     }
 }
 
 struct Worker<Context> {
     connection_pool: DieselPool,
     context: Context,
     job_registry: JobRegistry<Context>,
-    sender: SyncSender<Event>,
+    shutdown_when_queue_empty: bool,
+    poll_interval: Duration,
 }
 
 impl<Context: Clone + Send + 'static> Worker<Context> {
-    fn run_next_job(&self) {
-        let conn = &mut *match self.connection_pool.get() {
-            Ok(conn) => conn,
-            Err(e) => {
-                // TODO: Review error handling and possibly drop all usage of `let _ =` in this file
-                let _ = self.sender.send(Event::FailedToAcquireConnection(e));
-                return;
-            }
-        };
-
-        let job_run_result = conn.transaction::<_, diesel::result::Error, _>(|conn| {
-            debug!("Looking for next background worker job…");
-            let job = match storage::find_next_unlocked_job(conn).optional() {
-                Ok(Some(j)) => {
-                    let _ = self.sender.send(Event::Working);
-                    j
+    /// Run background jobs forever, or until the queue is empty if `shutdown_when_queue_empty` is set.
+    pub fn run(&self) {
+        loop {
+            match self.run_next_job() {
+                Ok(Some(_)) => {}
+                Ok(None) if self.shutdown_when_queue_empty => {
+                    debug!("No pending background worker jobs found. Shutting down the worker…");
```

Comment:
Maybe a `trace!` instead, just to limit ambient noise? (At a skim, I don't think we currently log in this case.)

Reply:
I tend to use `trace!` only for logging random, internal state variables and less for real log messages. Do you know of good guidelines on when to use what log level?

I'm think I'll keep the `debug!` for now since we only use `trace!` in one other place in the codebase.